### DISCUSSION TITLE
How to authorize two strategies & keep access tokens for both

### DISCUSSION THREAD

**Discussion by arikaa** [sergiodxa/remix-auth#188] [Q&A] [answered]
My application has a need to authenticate using two separate strategies and keep the access tokens around for both.

My current implementation looks like this:

```ts
// in auth.server.ts
export let authenticator = new Authenticator<User>(sessionStorage, {
  sessionKey: "sessionKey",
  sessionErrorKey: "sessionErrorKey",
});

let strategy1 = new Strategy({
  name: "strategy1"
  ...
  },
  async ({ accessToken, profile }) => {
    return { accessToken, profile };
  }
);
let strategy2 = new Strategy({
  name: "strategy2"
  ...
  },
  async ({ accessToken, profile }) => {
    return { accessToken, profile };
  });

authenticator.use(strategy1);
authenticator.use(strategy2);

// in root.tsx
export const loader: LoaderFunction = async ({ request }) => {
  const resultFromStrategy1 = await authenticator.authenticate(
    "strategy1",
    request
  );
// attempting to save session to a cookie to prevent override
  let headers = new Headers();
  if (resultFromStrategy1) {
    if (resultFromStrategy1?.accessToken) {
      const cookie = await myCookie.serialize(resultFromStrategy1?.accessToken);
      headers.append("Set-Cookie", cookie);
    }
  }
  const resultFromStrategy2 = await authenticator.authenticate(
    "strategy2",
    request
  );

  const authResponse = await authenticator.isAuthenticated(request);

  return json(
    {
      user: authResponse?.profile, // shows profile from resultFromStrategy1 which to me is unexpected
      accessToken: authResponse?.accessToken,
    },
    { headers }
  );
};
```

Is there a better way to accomplish this?

**Comment by sergiodxa** [comment]
If you store your access tokens in your DB, you could keep both around.

I'm not sure what your strategies does but two OAuth2-based strategies on the same request shouldn't work since you need to send the user to authorize your app to login with both providers.

**Reply by arikaa** [reply]
Thank you for the quick response! I might have to keep looking for a solution outside of storing in the database. In my application, I need to authenticate with Microsoft Graph to gain an access token for the logged in user and also for a custom API that is registered in Azure as well. Azure doesn't allow a single call to connect these since they connect with different _scopes_ to different _audiences_. Each needs a separate access token, which is why I was looking into how to support multiple.

**Comment by arikaa** [comment, marked as answer]
Following up incase others are trying to figure out an approach for the Microsoft case. I ended up calling fetch separately with a request to gain an access token for service-to-service authentication to validate API calls from my client to the backend. And I continued using remix-auth for the current logged in user.
